libc-top-half/musl/src/thread/pthread_create.c

```diff
@@ -364,6 +450,26 @@ int __pthread_create(pthread_t *restrict res, const pthread_attr_t *restrict att
 		if (ret)
 			__wait(&args->control, 0, 3, 0);
 	}
+#else
+	/* As in the unmodified version, all spawn failures translate to EAGAIN.
+	 * If the spawn did succeed, then wait for the TID to be populated by the
+	 * child thread. */
+	if (ret != 0) {
+		ret = -EAGAIN;
+	} else {
+		int timeout_ns = -1; // TODO
+		int wait_ret = __builtin_wasm_memory_atomic_wait32((int *)new->tid, 0, timeout_ns);
+		if (wait_ret == 1) {
+			// Loaded value != expected value; the pthread was not set up
+			// correctly.
+			return -EINVAL;
+		}
+		if (wait_ret == 2) {
+			// The timeout expired.
+			return -ETIMEDOUT;
+		}
+	}
```

Comment:
We can get rid of this `wait` (and the corresponding `notify` in `wasi_thread_start`) if we return the TID from `wasi_thread_spawn`.

Reply:
Presumably we can also get rid of the wait by using a purely userspace/fake (e.g. monotonically increasing atomic) tid?

Reply:
Yes, that would be another way to do this. In the discussion we had, others considered it necessary for the host to define this TID since it may need to maintain of mapping of that to real TIDs.

Reply:
Actually, thinking about this more, I now think a monotically increasing atomic TID controlled by the Wasm side (not the host) will still require a `wait`/`notify`: there are multiple threads that need the TID to continue &mdash the parent thread and the child thread. If the TID is incremented in the parent, then the child must be notified that the TID has been assigned before it continues to run the user function so that the TID is available to that code. If the TID is incremented in the child, then the parent must be notified of the new TID so that it can return with a correctly filled out `struct pthread`. Since both threads need the TID to continue, we would still need the `wait`/`notify`.

Reply:
To repeat what was just said out meeting, I think if you have a completely user-space TID it can be full created/decided prior the thread_spawn syscall (e.g. it can just be part of the pthread_t struct that is passed to the child).

Reply:
What ID will the main thread get? Is it going to be `0`, and the new TIDs start from `1`? Also, given that the host won't have access to those TIDs, what's the plan to implement (in the future) functions that manipulates native threads (e.g. setting priority)? Is the TID created in the userspace being passed to the `wasi_thread_spawn` so the runtime can keep track on the mapping?

Reply:
We need to decide if we want to bake in the idea that that host and the module should share a concept of TID from day 1.   With out initial goal of just creating `thread_spawn` there is no need to such a TID concept.

The need for way to pass a shared TID to the host will only exist once we add API such as `set_priority` or `thread_kill`.. which we don't have immediate plans for but can see the possible need for in the future.

I guess it comes down of two options:

1) add a shared TID concept from day today by adding a (currently useless) argument to `thread_spawn` or to the thread entry point function.
2) Add a new argument to one of those functions at the point when we add those new APIs. 
3) Add a new API that is something like `get_current_thread_id()` at the point when we add those new functions.

I think (3) has some downsides, but if we go with (1) then we can delay this discussion and focus just one what we need today for MVP.

Reply:
I *guess* sooner or later we'll need a host to know about the thread ID. Option 1. sounds reasonable; I'd rather stay away from 2. as I think we'd have to break the API to make that change.

Reply:
Yes. My comment was related to the case where the thread ID is generated in the userspace - in that case, we probably want to add another parameter to `wasi_thread_spawn` to tell the runtime the calculated thread id (so it can maintain the mapping).

Reply:
Yes I kind of like that option... since it means that user space never sees anything but its own TID which it generated.   

Then the host doesn't have to generate or share TIDs of its own and there is no possible figerprinting that can be done based on the host-provided TIDs.

Reply:
@sbc100 Can you say more about your reasons for wanting userspace to never see anything but its own TID? What kind of fingerprinting are you concerned about here?

Reply:
I was thinking that any information leaks from the host are a potential boon to the attacker, and TIDs could be of particular interest.

For example, I imagine that observing a particular pattern of TIDs could leak the identity of the host OS.  It doesn't seem like a major risk, just one that might be good to avoid if there are is an easy way to do so.

Making the TID (a) optional, and (b) created in user space (at the point of thread creation) , seems like a easy design choice that avoids host-side TIDs from being passed to the client module.  

What if we made the TID a (optional) second parameter to the thread creation function?:

```
i32: Thread ID that can be used in later system calls to identify the newly created thread.  0 means the thread cannot be referred to by such system calls.  Thread creation will fail if ID is already used by currently running thread.
```

Reply:
My thinking was that hosts where TIDs are sensitive can easily use a hash table or so to hide the host TIDs from wasm. Wasm-generated TIDs would ultimately need the exact same kind of hash table, to map wasm-generated TIDs to host TIDs, so it wouldn't be significantly different. And it would avoid having two different kinds of threads, with-TID and without-TID, which could potentially be an awkward special case in some settings.

Reply:
For later reference: the main reason for host-generated TIDs is that, in a scenario where multiple modules are linked together from different thread paradigms, it would be difficult to coordinate a central "atomically-incremented TID counter." E.g., if a Rust module and a C module both had Wasm-generated TIDs and were linked togethers, they would duplicate each other without some tricky toolchain coordination.